core/src/main/java/org/kohsuke/stapler/LocaleDrivenResourceProvider.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2019 Daniel Beck
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification, are permitted provided
+ * that the following conditions are met:
+ *
+ *     * Redistributions of source code must retain the above copyright notice, this list of
+ *       conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright notice, this list of
+ *       conditions and the following disclaimer in the documentation and/or other materials
+ *       provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS
+ * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
+ * AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
+ * OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package org.kohsuke.stapler;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ServiceLoader;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Service provider interface allowing to hook into webapp resource lookup.
+ *
+ * This cannot be made a property of WebApp as other behavior customizations, as webapp resource lookup is done before we have StaplerRequest/StaplerResponse.
+ */
+public abstract class LocaleDrivenResourceProvider {
+    /**
+     * Returns the URL corresponding to the specified resource path.
+     *
+     * The path can take two forms:
+     *
+     * <ul>
+     *     <li>A full URL represented as string. In Jenkins this is typically plugin webapp resources.</li>
+     *     <li>A path with leading slash. These are actual webapp resources, with that being the implicit base directory.</li>
```

Comment:
A potentially better design would be to have independent methods `lookupUrl` and `lookupPath`. Thoughts?

Reply:
Can't figure out the use cases. The interface method may not need to take care of those details.

Reply:
I'm generally in favor of using explicitly named separate methods when there is a clear separation of responsibility or implementation. In this case, though, I'm not sure that separation exists. Both forms are valid types of URLs.

Is there some reason we think the caller or implementor might need to handle these two forms differently or separately?

Reply:
> Is there some reason we think the caller or implementor might need to handle these two forms differently or separately?

It's most of what https://github.com/daniel-beck/localization-support-plugin/blob/master/src/main/java/io/jenkins/plugins/localization/support/stapler/LocaleDrivenResourceProviderImpl.java does.

"From a plugin or not" is the same distinction as the two variants documented here.

Reply:
> Both forms are valid types of URLs.

Note that the second one (plugin resources) is just `/path/on/file/system` pointing to `$JENKINS_HOME/plugins/<ID>/…`.